setuptools/dist.py

```diff
@@ -964,7 +966,28 @@ def run_command(self, command):
         # Postpone defaults until all explicit configuration is considered
         # (setup() args, config files, command line and plugins)
 
-        super().run_command(command)
+        with self._override_get_fullname():
```

Comment:
Hi @di, thank you very much for the update.

I have a question here (more out of curiosity really).

I am supposing that you needed to adopt this more complex approach because probably you found issues when trying a simpler one, e.g. something like the following:

```python
       def get_fullname(self):
            return "{}-{}".format(
                canonicalize_name(self.get_name()),
                canonicalize_version(self.get_version()),
            )
```

(i.e. simply overwrite via inheritance the method definition).

Do you remember which problems the simpler approach caused?

Reply:
@abravalheri Which class inherits from `DistributionMetadata` here that I would add that function to? I didn't see one, and I believe it gets instantiated within `distutils`, not within `setuptools`.

I'm following the pattern here: https://github.com/pypa/setuptools/blob/6ee23bf0579c52e1cbe7c97fc20fd085ff2a25c7/setuptools/command/sdist.py#L80-L99

But maybe I should do something like https://github.com/pypa/setuptools/blob/main/setuptools/monkey.py instead?

It's unclear to me what would be preferred.

Reply:
If I am not mistaken `get_fullname` is "dynamically" defined in the `Distribution` class via https://github.com/pypa/setuptools/blob/v69.2.0/setuptools/_distutils/dist.py#L151-L153. So I was supposing that we could just overwrite it in the `setuptools.dist.Distribution`.

But you are right, when the line https://github.com/pypa/setuptools/blob/v69.2.0/setuptools/dist.py#L307 runs, it gets overwritten anyway.

So yeah, now I can see what would be the problem.

Maybe we can do something like `self.get_fullname = self._get_fullname_canonicalized` just after the line https://github.com/pypa/setuptools/blob/v69.2.0/setuptools/dist.py#L307 (instead of the context manager)?

There is also the monkeypatch solution that you mentioned... Jason would be the best person to advise on that.

Reply:
Yes, that would work, although I don't really see any precedent for doing that. It also would only patch that instance, and not all `Distribution`s that might get instantiated, although it seems like that might not be an issue.

I'm happy to switch to the `monkey.py` solution instead if that makes more sense.

Reply:
My instinct is that the code should be first and foremost implemented in the canonical class. That is, what happens if we alter `distutils.dist.DistributionMetadata.get_fullname` to have the new behavior? Of course, that would need to be implemented in [pypa/distutils](/pypa/distutils), but since we maintain that, it should be straightforward to get that behavior ported here. That wouldn't fix the issue if distutils is used from stdlib, but that mode is soon to be deprecated, and if it needs to be supported, the `monkey` technique already provides forward ports of that functionality, so we should use that.

And if `get_fullname` is used elsewhere to mean non-canonicalized names, we can add a new method `get_filename` (or `get_canonical_full_name` or similar) and plumb that through.

Reply:
In cd57ac3f1, I explored simply altering get_fullname and the tests pass. That patch isn't viable because it alters `_distutils` (and also reaches back into setuptools for packaging), but it at least demonstrates that all the Setuptools tests pass. Next, I'll explore what happens if this change is introduced into distutils.

Reply:
> In https://github.com/pypa/setuptools/commit/cd57ac3f1c00479d70c1db8991bac5f3c0eda5df, I explored simply altering get_fullname and the tests pass. That patch isn't viable because it alters _distutils (and also reaches back into setuptools for packaging), but it at least demonstrates that all the Setuptools tests pass. Next, I'll explore what happens if this change is introduced into distutils.

Yes, this is what I had originally done in the first commit (b0135f5), which @abravalheri said wasn't viable.

Reply:
I applied a similar diff to the distutils project and most of the tests pass and the 5 failures look like intentional changes to align with the PEP:

```
FAILED distutils/tests/test_sdist.py::TestSDist::test_prune_file_list - AssertionError: assert ['fake-1.zip'] == ['fake-1.0.zip']
FAILED distutils/tests/test_sdist.py::TestSDist::test_make_distribution - AssertionError: assert ['fake-1.tar'...ake-1.tar.gz'] == ['fake-1.0.ta...e-1.0.tar.gz']
FAILED distutils/tests/test_sdist.py::TestSDist::test_add_defaults - AssertionError: assert ['fake-1.zip'] == ['fake-1.0.zip']
FAILED distutils/tests/test_sdist.py::TestSDist::test_manual_manifest - FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/f2/2plv6q2n7l932m2x004jlw340000gn/T/tmpa_mhw7wg/dist/fake-1.0.tar.gz'
FAILED distutils/tests/test_sdist.py::TestSDist::test_make_distribution_owner_group - FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/f2/2plv6q2n7l932m2x004jlw340000gn/T/tmpoypx9oc9/dist/fake-1.0.tar.gz'
```

So it seems to me the best approach will be to apply this change to distutils. A few things will need to happen before that can occur:

- distutils is currently ahead of Setuptools, so let's get those changes merged into Setuptools and released, so they're not conflated with this change
- distutils currently doesn't have access to anything in packaging. We'll want to vendor packaging or at least these functions from packaging to implement the canonicalization


Additionally, I'd like to see more tests in distutils covering expectations about these names (probably just docstrings on get_fullname covering more cases).

Reply:
In https://github.com/pypa/distutils/pull/244, I've proposed applying the changes in distutils.

Next, we'll want to adapt this change to provide a forward-port of that functionality in monkey.